File: Data Structures & Algorithms/buy-and-sell-crypto/submission-1.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        
        if(prices.size() <= 1){
            return 0;
        }

        int l = 0;
        int r = 1;
        int maxprofit = 0;

        while(r < prices.size()){
            maxprofit = max(maxprofit, prices.at(r) - prices.at(l));
            cout << l << " r: " << r << "\n";
            if(prices.at(l) >= prices.at(r)){
                l = r;
            }
            r++;
        }

        return maxprofit;
    }
};

File: Data Structures & Algorithms/buy-and-sell-crypto-with-cooldown/submission-1.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        memo = {}

        def opt(i, buying):

            if i >= len(prices):
                return 0
            
            if (i, buying) in memo:
                return memo[(i, buying)]
            
            res = opt(i + 1, buying)
            if buying:
                res = max(opt(i+1, not buying) - prices[i], res)
            else:
                res = max(opt(i + 2, not buying) + prices[i], res)
            
            memo[(i, buying)] = res

            return res
        
        return opt(0, True)
